scripts/pf2e-extempore-effects.js

```diff
@@ -179,31 +196,74 @@ Hooks.on('renderEffectsPanel', (panel, $html) => {
 
 // Wrap time-change functions so that the GM gets a notification on every Hidden condition that runs out of time
 const onUpdateWorldTime_Wrapper = (wrapped, ...args) => {
-  if (!game.user.isGM) return wrapped(...args)
-  const newWorldTime = args[0]
+  // if (!game.user.isGM) return wrapped(...args) // We have to move this check further down so player clocks stay in sync with GM clocks
+
+  const effectsWithDurations = game.pf2e.effectTracker.effects
+  if (!effectsWithDurations.length) return wrapped(...args);
+
   const oldWorldTime = game.time.worldTime
+  const willExpirationDeleteEffects = game.settings.get('pf2e', 'automation.removeExpiredEffects')
+  const effectNotificationSetting = game.settings.get(MODULE_ID, 'notifications-for-expired-effects')
+  const requestedNewWorldTime = args[0]
+
+  let newWorldTime = requestedNewWorldTime
+  let expireEffectMessageShown = false
+
+  // We don't want to go further in time than the next effect expiration or the requested time, which ever comes first
+  // Lets iterate through the effects and find the first one that will expire before the requested time, respecting the notification settings
+  if (game.settings.get(MODULE_ID, 'notifications-pause-for-expired-effects') && effectNotificationSetting !== 'disabled') {
+    let firstExpiringEffect = null;
+    for (const effect of effectsWithDurations) {
+      if (effect.isExpired) continue;
+      const isSecretEffect = effect.system.unidentified;
+      if (effectNotificationSetting === 'only_unidentified' && !isSecretEffect) continue;
+
+      if ((oldWorldTime + effect.remainingDuration.remaining) <= requestedNewWorldTime) {
+        firstExpiringEffect = effect;
+        break;
```

Comment:
I think this code is a bit wrong, because it breaks on the first matching effect that it finds in the list, rather than checking all effects and picking the one that expires the earliest.
This means if you have 2 effects in the list, and the second one expires earlier than the first one, when you advance time by a high amount it will skip to the expiration of the first one;  rather than skipping to the second one.

Reply:
Hmm, in my testing it always seemed like the effects return already sorted, earliest first?
But I could put a sort on that just to make sure...